### PR TITLE
Application boot and testing bootstrap changes

### PR DESCRIPTION
**Pull request by hypeJunction**
My last 5 cents and I really can't see Elgg code anymore.

### CODE REVIEW

File: engine/classes/ElggPlugin.php [hunk 1]
```diff
@@ -865,7 +865,10 @@ public function start($flags) {
 		// include start file if it exists
 		if ($flags & ELGG_PLUGIN_INCLUDE_START) {
 			if ($this->canReadFile('start.php')) {
-				require_once "{$this->path}/start.php";
+				$result = Includer::requireFileOnce("{$this->path}/start.php");
+				if ($result instanceof \Closure) {
+					$result(_elgg_services()->hooks->getEvents(), _elgg_services()->hooks);
```
Comment (mrclay) [inline engine/classes/ElggPlugin.php]: 👎  We shouldn't expose our internal event API objects to plugins.

Comment (mrclay) [inline engine/classes/ElggPlugin.php]: https://github.com/hypeJunction/Elgg/pull/1

File: engine/tests/classes/Elgg/IntegrationTestCase.php [hunk 2]
```diff
@@ -143,18 +118,26 @@ public static function tearDownAfterClass() {
...
-
+		$app = Application::getInstance();
 
-		$this->assertEquals($this->_testing_hooks, _elgg_services()->hooks->getAllHandlers(), "
-			The test has failed to clean up hook registrations after itself.
-		");
+		foreach ($this->_testing_hooks as $type => $hooks) {
+			foreach ($hooks as $hook => $handlers) {
+				foreach ($handlers as $handler) {
+					$this->assertTrue($app->_services->hooks->hasHandler($type, $hook, $handler));
```
Comment (hypeJunction) [inline engine/tests/classes/Elgg/IntegrationTestCase.php]: Does this make sense? I had an afterthought that this might be incorrect

Comment (mrclay) [inline engine/tests/classes/Elgg/IntegrationTestCase.php]: Why make assertions in tearDown?

Comment (hypeJunction) [inline engine/tests/classes/Elgg/IntegrationTestCase.php]: To make sure tests don't alter the state of hooks. I guess it's redundant now that we start a new application for each test

Comment (mrclay) [inline engine/tests/classes/Elgg/IntegrationTestCase.php]: I suppose we could directly compare the output of getAllHandlers() afterwards? I don't know how strict that needs to be.

Comment (hypeJunction) [inline engine/tests/classes/Elgg/IntegrationTestCase.php]: I tried that and there are some hooks that use closures with injected session and various timestamps, so it doesn't work

File: engine/classes/Elgg/Application.php [hunk 3]
```diff
@@ -772,7 +772,19 @@ public static function test() {
 		self::loadCore();
 
 		$app = IntegrationTestCase::createApplication();
-		if (!$app) {
+		if ($app) {
+			// To speed up database testing, we assign __testing metadata to all created entities
+			// We delete them at shutdown
+			register_shutdown_function(function() {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Application.php]: This is hacky but will have to do for now

**Comment by hypeJunction** [comment]
I don't know what's up with the HHVM build, and why the tests are taking forever to complete.

**Comment by hypeJunction** [comment]
Flushing the cache for each test seems to be culprit.

**Comment by mrclay** [comment]
#11185 Suggesting we drop HHVM.

**Comment by hypeJunction** [comment]
For some reason memcached is enabled on HHVM builds

**Comment by hypeJunction** [comment]
@mrclay I've done all I could here. Something is throwing somewhere, but not everywhere, and I don't have what to do with it. Take a look to see if you see something I don't see.

**Comment by hypeJunction** [comment]
Perhaps new PHPUnit is stricter about time execution limit https://phpunit.de/manual/current/en/risky-tests.html. I see it to 0 in bootstrap.php, but it doesn't seem to be respected

**Comment by hypeJunction** [comment]
Replaced by #11188

**Pull request closed without merging**